### PR TITLE
[Saitosan] Add new extractor

### LINKED ISSUE #224: [Site Request] http://saitosan.net/
## Checklist

- [x] I'm reporting a new site support request
- [x] I've verified that I'm running yt-dlp version **2021.04.03**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones

## Example URLs

- http://saitosan.net/broadcast.html

## Description

Japanese live website.

### PR DESCRIPTION
**Pull request by llacb47**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fixes #224

I decided to try implementing this extractor since I was bored. The code is kinda terrible since it parses socket.io messages itself without relying on a dedicated socket.io client library like [this](https://github.com/miguelgrinberg/python-socketio). I don't even speak Japanese and have basically no idea what this website is lol. Also, the broadcasts last for a very short time and that makes testing very difficult, and the broadcasts have very little metadata to extract. Let me know if I should remove the uploader field since as far as I can tell it's always the same as the title. 

`https://www.yudo.jp/en/service/saitosan`

### CODE REVIEW

**Comment by llacb47** [comment]
By the way you can find live broadcasts to test with at `http://www.saitosan.net/broadcast.html` (doesn't work with HTTPS).

File: yt_dlp/extractor/saitosan.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        def format_socket_response_as_json(data):
+            return self._parse_json(data[data.find('{'):data.rfind('}') + 1], b_id)
+
+        base = 'http://hankachi.saitosan-api.net:8002/socket.io/?transport=polling&EIO=3'
+        sid = format_socket_response_as_json(self._download_webpage(base, b_id, note='Opening socket')).get('sid')
+        base += "&sid=" + sid
+
+        self._download_webpage(base, b_id, note="Polling socket")
+        payload = '420["room_start_join",{"room_id":"' + str(b_id) + '"}]'
+        payload = str(len(payload)) + ":" + payload
```
Comment (pukkandan) [inline yt_dlp/extractor/saitosan.py:55]: Using `%` or `format` is preferred over `+`

File: yt_dlp/extractor/saitosan.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        self._download_webpage(base, b_id, data=payload, note="Polling socket")
```
Comment (pukkandan) [inline yt_dlp/extractor/saitosan.py:57]:
```suggestion
        self._download_webpage(base, b_id, data=payload, note='Polling socket with payload')
```

Also use `'` everywhere (unless the string has `'` ofc) for consistency across the whole project

File: yt_dlp/extractor/saitosan.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        should_continue = format_socket_response_as_json(self._download_webpage(base, b_id, note="Polling socket")).get('ok')
+        if not should_continue:
+            # The socket does not give any specific error messages.
+            self.raise_no_formats("The socket reported that the broadcast could not be joined. Maybe it's offline or the URL is incorrect", expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/saitosan.py:61]:
```suggestion
            raise ExtractorError(
                'The socket reported that the broadcast could not be joined. Maybe it's offline or the URL is incorrect',
                expected=True, video_id=b_id)
```
`raise_no_formats` is for when metadata is available, but download links are not

File: yt_dlp/extractor/saitosan.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+
+class SaitosanIE(InfoExtractor):
+    IE_NAME = 'Saitosan'
```
Comment (pukkandan) [inline yt_dlp/extractor/saitosan.py:8]: It is not necessary to define `IE_NAME` when it is same as the class name. Ofc, no harm in defining it either

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
> I decided to try implementing this extractor since I was bored ... I don't even speak Japanese and have basically no idea what this website is lol. 

Thanks! Not many people put in the effort to fix, much less add a new extractor that they themselves don't use (I am guilty of this too)

File: yt_dlp/extractor/saitosan.py [hunk 5]
```diff
@@ -58,7 +58,9 @@ def format_socket_response_as_json(data):
         should_continue = format_socket_response_as_json(self._download_webpage(base, b_id, note="Polling socket")).get('ok')
         if not should_continue:
             # The socket does not give any specific error messages.
-            self.raise_no_formats("The socket reported that the broadcast could not be joined. Maybe it's offline or the URL is incorrect", expected=True)
+            raise ExtractorError(
+                'The socket reported that the broadcast could not be joined. Maybe it's offline or the URL is incorrect',
```
Comment (pukkandan) [inline yt_dlp/extractor/saitosan.py:62]:
```suggestion
                "The socket reported that the broadcast could not be joined. Maybe it's offline or the URL is incorrect",
```

**Comment by llacb47** [comment]
Now I'm confused, why did you change the single quotes to double quotes? https://github.com/yt-dlp/yt-dlp/pull/350/commits/ea6e081aad55e42f32e79b3938894cf8a218ebbf

File: yt_dlp/extractor/saitosan.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        self._download_webpage(base, b_id, data=payload, note='Polling socket with payload')
+        should_continue = format_socket_response_as_json(self._download_webpage(base, b_id, note="Polling socket")).get('ok')
+        if not should_continue:
+            # The socket does not give any specific error messages.
+            raise ExtractorError(
+                "The socket reported that the broadcast could not be joined. Maybe it's offline or the URL is incorrect",
+                expected=True, video_id=b_id)
+
+        self._download_webpage(base, b_id, data='26:421["room_finish_join",{}]', note="Polling socket")
```
Comment (llacb47) [inline yt_dlp/extractor/saitosan.py:65]: Double quotes here...

File: yt_dlp/extractor/saitosan.py (hunk 5 above)
Comment (llacb47) [reply yt_dlp/extractor/saitosan.py]: Oh I realize now. Whoops.

File: yt_dlp/extractor/saitosan.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+        def format_socket_response_as_json(data):
+            return self._parse_json(data[data.find('{'):data.rfind('}') + 1], b_id)
+
+        base = 'http://hankachi.saitosan-api.net:8002/socket.io/?transport=polling&EIO=3'
+        sid = format_socket_response_as_json(self._download_webpage(base, b_id, note='Opening socket')).get('sid')
+        base += '&sid=' + sid
+
+        self._download_webpage(base, b_id, note='Polling socket')
+        payload = '420["room_start_join",{"room_id":"' + str(b_id) + '"}]'
```
Comment (pukkandan) [inline yt_dlp/extractor/saitosan.py:54]:
```suggestion
        payload = '420["room_start_join",{"room_id":"%s"}]' % b_id
```

**Comment by nao20010128nao** [comment]
As a Japanese speaker, I think it's okay to make `uploader` field same as `title`. They don't give us much info enough.

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/saitosan.py (hunk 7 above)
Comment (llacb47) [reply yt_dlp/extractor/saitosan.py]: `compat_str` necessary or not anymore since Python 2 support was dropped?

**Comment by llacb47** [comment]
> 
> 
> As a Japanese speaker, I think it's okay to make `uploader` field same as `title`. They don't give us too much info.

So keep it or delete it?

**Comment by llacb47** [comment]
Otherwise lgtm?

**Comment by nao20010128nao** [comment]
> So keep it or delete it?

I have no idea. @pukkandan

**Comment by pukkandan** [comment]
The PR is good as-is. The only reason it isn't merged is coz I didn't get time to test it yet

**Pull request merged**